Fix memory leak in linux demangler

Manu343726 opened the pull request:
The __cxa_demangle() call was leaking the demangled name buffer
(Confirmed by valgrind).

If I understood the __cxa_demangle() docs right, the issue was that you were
mixing the two operating modes (User buffer vs __cxa_demangle()
allocated buffer).
You passed a garbage pointer to __cxa_demangle(), which expects null
or a pointer to a malloc() allocated block of _size_ bytes. Note also size is
0 in your implementation. Basically the leak comes from __cxa_demangle() doing a
realloc() on your garbage pointer.

The fix is as simple as telling __cxa_demangle() to allocate the buffer
itself.

Just for reference, here is the __cxa_demangle() documentation
explaining this behavior:

https://gcc.gnu.org/onlinedocs/libstdc++/libstdc++-html-USERS-4.3/a01696.html

Manu343726 commented:
Here's an example of valgrind output: https://gist.github.com/Manu343726/24453dfcc6845c09475e16579a32a7ab

bombela commented:
`_demangle_buffer.release()` releases the held pointer back (ie: doesn't manage it anymore). `abi::__cxa_demangle` can then `realloc` it as needed. The returned pointer ownership is then taken back into `_demangle_buffer` via `_demangle_buffer.reset()`.

```c++
_demangle_buffer.reset(
		abi::__cxa_demangle(funcname, _demangle_buffer.release(),
			&_demangle_buffer_length, 0)
		);

```

This works assuming that after initialization, ` _demangle_buffer.release()` returns a NULL pointer.
We can verify this (https://github.com/Manu343726/backward-cpp/blob/ad84214192f09ef763cde5e66cd0aeeb63ee6f4d/backward.hpp#L329):
```c++
	explicit handle(): _val(), _empty(true) {}
```

The documentation of `char* abi::__cxa_demangle` says:

> output_buffer: A region of memory, allocated with malloc, of *length bytes, into which the demangled name is stored. If output_buffer is not long enough, it is expanded using realloc. output_buffer may instead be NULL; in that case, the demangled name is placed in a region of memory allocated with malloc.


> length: If length is non-NULL, the length of the buffer containing the demangled name is placed in *length.

```c++
char* abi::__cxa_demangle	(	const char * 	mangled_name,
    char*    output_buffer,
    size_t*  length,
    int*     status	 
)	
```

So it should be possible to feed `output_buffer==NULL` to the first call. `abi::__cxa_demangle` will `malloc` a buffer and store update `*length`. The following calls, it should be possible to feed back the previous `output_buffer` pointer as well as `length`. Then if `*length` is not enough, `abi::__cxa_demangle` should `realloc` the `output_buffer` and update `*length`.

Maybe `abi::__cxa_demangle` is broken, maybe my code is not initializing the initial `_demangle_buffer` pointer to NULL. That's something to investigate.

I would like to try reusing the same buffer as much as possible, avoiding extra calls to malloc if not needed.

Manu343726 commented:
Hi,

After taking time to read your comments in depth, I reverted my changes and added a Travis CI build to check if the usage of backward-cpp (with your unit tests) was really leaking memory. As you can see [here](https://travis-ci.org/Manu343726/backward-cpp/builds/182987533) it seems that there are no problems. I will try to reproduce my use case in one of your tests and see if it stills leaks.

Manu343726 commented:
Oh, If you mind you can merge that commit with the CI build ;)

bombela commented on conanfile.py:
```diff
@@ -0,0 +1,57 @@
+class BackwardCpp(ConanFile):
+    settings = 'os', 'compiler', 'build_type', 'arch'
+    name = 'backward'
+    url = 'https:77github.com/Manu343726/backward-cpp'
```
shouldn't the URL be  "https://github.com/bombela/backward-cpp" ? And what about the 77 instead of // ?

bombela commented on test_package/conanfile.py:
```diff
@@ -0,0 +1,15 @@
+
+class TestBackward(ConanFile):
+    settings = 'os', 'compiler', 'build_type', 'arch'
+    requires = 'cmake-utils/0.0.0@Manu343726/testing', 'backward/0.0.0@Manu343726/testing'
```
URL thingy?

bombela commented:
Nevermind I realized you offered the CI commit for merge.

bombela commented:
Merged the CI commit.

Manu343726 commented:
Sorry, I followed working on the conan support in my upstream without realizing that commits will appear here. Anyway, I will be glad to PR again and integrate backward-cpp officially in conan. It should be as easy as creating your own account, merge the scripts, and own the package. Of course ask me if you're interested and need help on the task.

bombela commented:
Thank you for the proposition yet I would prefer somebody who knows and use
Conan to maintain the package, not me. If you want to maintain it let me
know and I will give you write access to the repo.

On Sun, Dec 18, 2016, 08:54 Manu Sánchez <notifications@github.com> wrote:

> Sorry, I followed working on the conan support in my upstream without
> realizing that commits will appear here. Anyway, I will be glad to PR again
> and integrate backward-cpp officially in conan. It should be as easy as
> creating your own account, merge the scripts, and own the package. Of
> course ask me if you're interested and need help on the task.
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/bombela/backward-cpp/pull/38#issuecomment-267832252>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANMjBFektj1T5Umkm97rfcxCXTXgqXdks5rJWVKgaJpZM4Jr4NI>
> .
>

Manu343726 commented:
Great, I would happily maintain it for you. Thanks!

bombela commented:
Done :)

Manu343726 commented:
Hi again,

I'm cleaning up things so I will only have to push one simple commit into your master branch. Before doing so, I want to know if you have a travis ci account which to point the CI build to (My commit includes a .travis.yml file to run your unit tests and then test the package). The build is currently using my account, but  as far I know we need an account of yours so builds are triggered when pushing to this repository.

BTW, you can check the commit [here](https://github.com/Manu343726/backward-cpp).

bombela commented:
this should do it: https://travis-ci.org/bombela/backward-cpp/

Manu343726 commented:
Great, there's only a minor detail left: The packaged currently hosted in conan.io is labeled as v0.0.0 meaning it points to the "latest" version in `master` (which is not exactly true). I did it that way since I thought v1.0.0 release was a bit old compared to the upstream. What do you think? Should I upload a 1.0.0 package pointing to the 1.0.0 release instead?

bombela commented:
yeah that sounds like a more reasonable solution. Note the latest version is 1.2, and its out of date. Feel free to add a new v1.3 tag if needed.

Manu343726 commented:
Ok. Then I'm going to:

 - Update CI badges in my commit to point to your builds.
 - Push the commit to your master branch.
 - Tag and release a v1.3 version with this latest changes.

Thanks for all the feedback.

bombela commented:
Sounds perfect! Thank you so much for doing all this!

On Sun, Dec 18, 2016 at 1:58 PM Manu Sánchez <notifications@github.com>
wrote:

> Ok. Then I'm going to:
>
>    - Update CI badges in my commit to point to your builds.
>    - Push the commit to your master branch.
>    - Tag and release a v1.3 version with this latest changes.
>
> Thanks for all the feedback.
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/bombela/backward-cpp/pull/38#issuecomment-267849869>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANMjLaA_WDd5d82wpWbedsdBjah58FJks5rJayEgaJpZM4Jr4NI>
> .
>
